File: Algorithms/p309_Best_Time_to_Buy_and_Sell_Stock_with_Cooldown.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        l = len(prices)
        if l<=1:
            return 0
        s0 = [None]*l # not holding
        s1 = [None]*l # holding
        s2 = [None]*l # just sold
        s0[0] = 0
        s1[0] = -prices[0]
        for i in range(1,l):
            s0[i] = max(s0[i-1],s2[i-1]) if i>1 else s0[0]
            s1[i] = max(s0[i-1]-prices[i],s1[i-1])
            s2[i] = s1[i-1]+prices[i]
        return max(s0[l-1],s2[l-1])

File: Algorithms/p309_Best_Time_to_Buy_and_Sell_Stock_with_Cooldown/p309_Best_Time_to_Buy_and_Sell_Stock_with_Cooldown.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int l = prices.size();
        if(l<=1) return 0;
        vector<int> s0(l); // not holding
        vector<int> s1(l); // holding
        vector<int> s2(l); // just sold
        s0[0] = 0;
        s1[0] = -prices[0];
        for(int i=1;i<l;++i){
            s0[i] = i>1 ? max(s0[i-1],s2[i-1]) : s0[0];
            s1[i] = max(s0[i-1]-prices[i],s1[i-1]);
            s2[i] = s1[i-1]+prices[i];
        }
        return max(s0[l-1],s2[l-1]);
    }
};
